File: src/sica_mapping/data/tables.py
"""Table builders for the West End map UI."""

from __future__ import annotations

import numpy as np
import pandas as pd
from html import escape


def buildings_table(pts_df: pd.DataFrame) -> pd.DataFrame:
    tbl = pts_df[
        [
            "b_id",
            "address",
            "local_area",
            "block_id",
            "units",
            "member_count",
            "member_share_building",
            "year_built",
            "owner_group",
            "owner_key",
            "member_count_all",
            "value_land",
            "value_bldg",
            "bldg_land_ratio",
            "has_vtu_member",
            "latest_membership_year",
            "housing_type",
            "rezoning_status",
            "rezoning_status_group",
        ]
    ].copy()
    tbl["member_share_pct"] = (tbl["member_share_building"] * 100).round(0).astype(int)
    tbl["source"] = "building"
    return tbl.drop(columns=["member_share_building"]).sort_values(
        ["member_count", "units"], ascending=[False, False]
    )


def _avg_units_per_bldg(total_units: pd.Series, buildings: pd.Series) -> pd.Series:
    return np.where(buildings > 0, total_units / buildings, 0.0).round(1)


def blocks_table(blocks_merged: pd.DataFrame) -> pd.DataFrame:
    tbl = blocks_merged.copy()
    tbl["median_year_built"] = tbl["median_year_built"].round().astype("Int64")
    tbl["avg_units_per_bldg"] = _avg_units_per_bldg(tbl["total_units"], tbl["buildings"])
    # Blocks with no buildings in this dataset fall back to "(Unknown)" (see
    # assign_block_labels); sort those last rather than first so a viewer sees
    # the actually-labeled, meaningful blocks before the empty-block bucket.
    tbl["_is_unknown"] = tbl["block_label"].str.startswith("(Unknown)")
    return tbl[
        [
            "block_id",
            "block_label",
            "local_area",
            "buildings",
            "total_units",
            "avg_units_per_bldg",
            "median_year_built",
            "member_buildings",
            "_is_unknown",
        ]
    ].sort_values(["_is_unknown", "block_label"], ascending=[True, True]).drop(
        columns=["_is_unknown"]
    )


def landlords_table(pts_df: pd.DataFrame) -> pd.DataFrame:
    owners_df = pts_df.copy()
    owners_df["owner_group"] = owners_df["owner_group"].fillna("(Unknown)")
    owners_df["owner_key"] = owners_df["owner_key"].fillna("unknown")
    landlords = (
        owners_df.groupby(["owner_group", "owner_key"])
        .agg(
            buildings=("address", "count"),
            total_units=("units", "sum"),
            member_buildings=("has_vtu_member", "sum"),
        )
        .reset_index()
    )
    landlords["avg_units_per_bldg"] = _avg_units_per_bldg(
        landlords["total_units"], landlords["buildings"]
    )
    return landlords.sort_values(["total_units", "buildings"], ascending=[False, False])


def neighbourhoods_table(pts_df: pd.DataFrame) -> pd.DataFrame:
    df = pts_df.copy()
    df["local_area"] = df["local_area"].fillna("(Unknown)")
    tbl = (
        df.groupby("local_area")
        .agg(
            buildings=("address", "count"),
            total_units=("units", "sum"),
            member_buildings=("has_vtu_member", "sum"),
        )
        .reset_index()
    )
    tbl["avg_units_per_bldg"] = _avg_units_per_bldg(tbl["total_units"], tbl["buildings"])
    return tbl.sort_values(["total_units", "buildings"], ascending=[False, False])


def rows_buildings(df: pd.DataFrame) -> str:
    rows = []
    for r in df.itertuples(index=False):
        block_val = "" if pd.isna(r.block_id) else int(r.block_id)
        units_val = "" if pd.isna(r.units) else int(r.units)
        year_val = "" if pd.isna(r.year_built) else int(r.year_built)
        has_vtu_val = 1 if bool(r.has_vtu_member) else 0
        membership_year_val = (
            "" if pd.isna(r.latest_membership_year) else int(r.latest_membership_year)
        )
        val_land = "" if pd.isna(r.value_land) else int(round(r.value_land))
        val_bldg = "" if pd.isna(r.value_bldg) else int(round(r.value_bldg))
        ratio_val = (
            "" if pd.isna(r.bldg_land_ratio) else round(float(r.bldg_land_ratio), 3)
        )
        member_total = (
            int(r.member_count_all)
            if hasattr(r, "member_count_all")
            else int(r.member_count)
        )
        bid = int(r.b_id)
        owner_key = escape(str(r.owner_key))
        local_area = "" if pd.isna(r.local_area) else str(r.local_area)
        housing_type = "" if pd.isna(r.housing_type) else str(r.housing_type)
        rezoning_status = "" if pd.isna(r.rezoning_status) else str(r.rezoning_status)
        rezoning_status_group = (
            "" if pd.isna(r.rezoning_status_group) else str(r.rezoning_status_group)
        )
        search_terms = " ".join(
            str(val).lower()
            for val in (
                r.address,
                local_area,
                block_val,
                units_val,
                int(r.member_count),
                r.owner_group,
                housing_type,
                rezoning_status,
            )
            if val not in ("", None)
        )
        search_attr = escape(search_terms)
        rows.append(
            (
                f'<tr data-bid="{bid}" data-owner="{owner_key}" data-block="{block_val}" '
                f'data-area="{escape(local_area)}" '
                f'data-value-land="{val_land}" data-value-bldg="{val_bldg}" '
                f'data-value-ratio="{ratio_val}" data-units="{units_val}" '
                f'data-year-built="{year_val}" '
                f'data-has-vtu-member="{has_vtu_val}" '
                f'data-latest-membership-year="{membership_year_val}" '
                f'data-member-total="{member_total}" data-search="{search_attr}" '
                f'data-source="building" data-synthetic="0" '
                f'data-housing-type="{escape(housing_type)}" '
                f'data-rezoning-status="{escape(rezoning_status)}" '
                f'data-rezoning-status-group="{escape(rezoning_status_group)}">'
                f'<td class="select-cell"><input type="checkbox" class="row-select" '
                f'data-type="building" data-target="{bid}"></td>'
                f"<td>{escape(str(r.address))}</td>"
                f'<td data-sort-value="{escape(local_area)}">{escape(local_area)}</td>'
                f'<td data-sort-value="{block_val}">{block_val}</td>'
                f'<td data-sort-value="{units_val}">{units_val}</td>'
                f'<td data-sort-value="{int(r.member_count)}">{int(r.member_count)}</td>'
                f'<td data-sort-value="{int(r.member_share_pct)}">{int(r.member_share_pct)}%</td>'
                f"<td>{escape(str(r.owner_group))}</td>"
                f'<td data-sort-value="{year_val}">{year_val}</td>'
                f'<td data-sort-value="{escape(housing_type)}">{escape(housing_type)}</td>'
                f'<td data-sort-value="{escape(rezoning_status)}">{escape(rezoning_status)}</td>'
                f"</tr>"
            )
        )
    return "\n".join(rows)


def rows_synthetic(records: list[dict]) -> str:
    """Unmatched SRO/co-op/rezoning records rendered as Buildings-table rows,
    flagged by source, with real-building-only fields left blank. See
    data.overlays.match_overlays for how these records are produced.
    """
    rows = []
    for rec in records:
        sid = escape(str(rec["synthetic_id"]))
        source = escape(str(rec.get("source") or ""))
        local_area = str(rec.get("local_area") or "")
        address = str(rec.get("address") or "")
        housing_type = str(rec.get("housing_type") or "")
        rezoning_status = str(rec.get("rezoning_status") or "")
        rezoning_status_group = str(rec.get("rezoning_status_group") or "")
        search_terms = " ".join(
            str(val).lower()
            for val in (address, local_area, source)
            if val not in ("", None)
        )
        rows.append(
            (
                f'<tr data-bid="{sid}" data-owner="" data-block="" '
                f'data-area="{escape(local_area)}" '
                f'data-value-land="" data-value-bldg="" data-value-ratio="" '
                f'data-units="" data-year-built="" data-has-vtu-member="0" '
                f'data-latest-membership-year="" data-member-total="0" '
                f'data-search="{escape(search_terms)}" '
                f'data-source="{source}" data-synthetic="1" '
                f'data-housing-type="{escape(housing_type)}" '
                f'data-rezoning-status="{escape(rezoning_status)}" '
                f'data-rezoning-status-group="{escape(rezoning_status_group)}">'
                f'<td class="select-cell"><input type="checkbox" class="row-select" '
                f'data-type="building" data-target="{sid}"></td>'
                f"<td>{escape(address)} <em>(unmatched {source})</em></td>"
                f'<td data-sort-value="{escape(local_area)}">{escape(local_area)}</td>'
                f"<td></td><td></td><td></td><td></td><td></td><td></td>"
                f'<td data-sort-value="{escape(housing_type)}">{escape(housing_type)}</td>'
                f'<td data-sort-value="{escape(rezoning_status)}">{escape(rezoning_status)}</td>'
                f"</tr>"
            )
        )
    return "\n".join(rows)


def rows_blocks(df: pd.DataFrame) -> str:
    rows = []
    for r in df.itertuples(index=False):
        block_id = int(r.block_id)
        block_label = escape(str(r.block_label))
        local_area = "" if pd.isna(r.local_area) else str(r.local_area)
        year_val = "" if pd.isna(r.median_year_built) else int(r.median_year_built)
        avg_units = float(r.avg_units_per_bldg)
        rows.append(
            f'<tr data-block="{block_id}" data-area="{escape(local_area)}" '
            f'data-bldgs="{int(r.buildings)}" data-units="{int(r.total_units)}" '
            f'data-vtu-bldgs="{int(r.member_buildings)}">'  # block
            f'<td class="select-cell"><input type="checkbox" class="row-select" '
            f'data-type="block" data-target="{block_id}"></td>'
            f'<td data-sort-value="{block_label}">{block_label}</td>'
            f'<td data-sort-value="{int(r.buildings)}">{int(r.buildings)}</td>'
            f'<td data-sort-value="{int(r.total_units)}">{int(r.total_units)}</td>'
            f'<td data-sort-value="{avg_units}">{avg_units:.1f}</td>'
            f'<td data-sort-value="{year_val}">{year_val}</td>'
            f'<td data-sort-value="{int(r.member_buildings)}">{int(r.member_buildings)}</td>'
            f"</tr>"
        )
    return "\n".join(rows)


def rows_landlords(df: pd.DataFrame) -> str:
    rows = []
    for r in df.itertuples(index=False):
        owner_key = escape(str(r.owner_key))
        units_val = "" if pd.isna(r.total_units) else int(r.total_units)
        avg_units = float(r.avg_units_per_bldg)
        rows.append(
            f'<tr data-owner="{owner_key}" '
            f'data-bldgs="{int(r.buildings)}" data-units="{units_val if units_val != "" else 0}" '
            f'data-vtu-bldgs="{int(r.member_buildings)}">'  # owner
            f'<td class="select-cell"><input type="checkbox" class="row-select" '
            f'data-type="owner" data-target="{owner_key}"></td>'
            f"<td>{escape(str(r.owner_group))}</td>"
            f'<td data-sort-value="{int(r.buildings)}">{int(r.buildings)}</td>'
            f'<td data-sort-value="{units_val}">{units_val}</td>'
            f'<td data-sort-value="{avg_units}">{avg_units:.1f}</td>'
            f'<td data-sort-value="{int(r.member_buildings)}">{int(r.member_buildings)}</td>'
            f"</tr>"
        )
    return "\n".join(rows)


def rows_neighbourhoods(df: pd.DataFrame) -> str:
    rows = []
    for r in df.itertuples(index=False):
        local_area = escape(str(r.local_area))
        units_val = "" if pd.isna(r.total_units) else int(r.total_units)
        avg_units = float(r.avg_units_per_bldg)
        rows.append(
            f'<tr data-area="{local_area}" '
            f'data-bldgs="{int(r.buildings)}" data-units="{units_val if units_val != "" else 0}" '
            f'data-vtu-bldgs="{int(r.member_buildings)}">'  # neighbourhood
            f'<td class="select-cell"><input type="checkbox" class="row-select" '
            f'data-type="neighbourhood" data-target="{local_area}"></td>'
            f"<td>{local_area}</td>"
            f'<td data-sort-value="{int(r.buildings)}">{int(r.buildings)}</td>'
            f'<td data-sort-value="{units_val}">{units_val}</td>'
            f'<td data-sort-value="{avg_units}">{avg_units:.1f}</td>'
            f'<td data-sort-value="{int(r.member_buildings)}">{int(r.member_buildings)}</td>'
            f"</tr>"
        )
    return "\n".join(rows)

File: frontend/src/tables.ts
/**
 * Sidebar table rows, ported from sica_mapping/data/tables.py. Attributes,
 * columns, sort keys and rounding follow the Python, because wiring.js reads
 * these rows as its filter model: data-* attributes, fixed column indices
 * (wiring.js cacheRowCells) and .row-select checkboxes. Known cosmetic
 * differences, neither affecting sorting or filtering: numeric sort values
 * render as JS numbers (15, not Python's 15.0), and tied rows may order
 * differently. Must render before wiring.js starts; it queries the rows once.
 */
import { formatAddress } from './address';
import { escapeHtml, isMissing, roundHalfEven, sameName } from './html';
import { EXPLAINERS, networkProvenance, ownerRowProvenance, provenanceIcon, tipPlacement } from './provenance';
import type { BlocksCollection, BuildingData, BuildingRecord } from './types';

const num = (v: unknown): number | null => (typeof v === 'number' && Number.isFinite(v) ? v : null);
/** Python int() of a present number, else "". */
const intStr = (v: unknown): string => {
  const n = num(v);
  return n === null ? '' : String(Math.trunc(n));
};
const text = (v: unknown): string => (isMissing(v) ? '' : String(v));

/** pandas sort_values(ascending=False): larger first, missing last. Stable. */
function descMissingLast(a: number | null, b: number | null): number {
  if (a === null && b === null) return 0;
  if (a === null) return 1;
  if (b === null) return -1;
  return b - a;
}

/**
 * A building with no known landlord. The export still writes the "unknown"
 * key; it is shown as "Not on record" and never ranked as a landlord.
 */
const notOnRecord = (key: unknown): boolean => isMissing(key) || key === 'unknown';

/** Rows in their current order, with data-pin="last" rows moved to the end (wiring.js sorting). */
export function pinLast<T extends { getAttribute(name: string): string | null }>(rows: T[]): T[] {
  const pinned = (r: T) => r.getAttribute('data-pin') === 'last';
  return [...rows.filter((r) => !pinned(r)), ...rows.filter(pinned)];
}

/** Python's str ordering (code points), not localeCompare. */
function byCodePoint(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

export function buildingRow(r: BuildingRecord): string {
  const bid = Math.trunc(Number(r.b_id));
  const block = intStr(r.block_id);
  const units = intStr(r.units);
  const year = intStr(r.year_built);
  const nIssues = intStr(r.n_issues);
  const land = num(r.value_land);
  const bldg = num(r.value_bldg);
  const ratio = num(r.bldg_land_ratio);
  const valLand = land === null ? '' : String(roundHalfEven(land));
  const valBldg = bldg === null ? '' : String(roundHalfEven(bldg));
  const ratioVal = ratio === null ? '' : String(roundHalfEven(ratio, 3));
  const area = text(r.local_area);
  const housing = text(r.housing_type);
  const onRecord = !notOnRecord(r.owner_key);
  const owner = onRecord ? text(r.owner_name) : '';
  const networkName = onRecord && !notOnRecord(r.network_key) ? text(r.network_name) : '';
  const network = sameName(networkName, owner) ? '' : networkName;
  const manager = text(r.managed_by);
  const sub = (line: string) => (line ? `<div class="cell-sub">${escapeHtml(line)}</div>` : '');
  const managerLine = sub(manager ? `Managed by ${manager}` : '');
  const landlordCell = onRecord
    ? `<td>${escapeHtml(owner)}${sub(network ? `Group: ${network}` : '')}${managerLine}</td>`
    : `<td class="not-on-record">Not on record${managerLine}</td>`;
  const names = [text(r.building_name), ...(Array.isArray(r.other_names) ? r.other_names.map(text) : [])];
  const search = [text(r.address), ...names, area, units, owner, network, manager, housing]
    .filter((v) => v !== '')
    .map((v) => v.toLowerCase())
    .join(' ');
  const a = escapeHtml(area);
  const h = escapeHtml(housing);
  const chinatown = r.in_chinatown === true ? '1' : '';
  const village = r.in_village_plan === true ? '1' : '';
  return (
    `<tr data-bid="${bid}" data-owner="${escapeHtml(text(r.owner_key))}" ` +
    `data-network="${escapeHtml(text(r.network_key))}" data-block="${block}" ` +
    `data-area="${a}" data-chinatown="${chinatown}" data-village="${village}" ` +
    `data-value-land="${valLand}" data-value-bldg="${valBldg}" ` +
    `data-value-ratio="${ratioVal}" data-units="${units}" ` +
    `data-year-built="${year}" data-search="${escapeHtml(search)}" ` +
    `data-housing-type="${h}" data-n-issues="${nIssues}" data-has-owner="${onRecord ? '1' : ''}">` +
    `<td class="select-cell"><input type="checkbox" class="row-select" ` +
    `data-type="building" data-target="${bid}"></td>` +
    `<td>${escapeHtml(formatAddress(r.address))}</td>` +
    `<td>${escapeHtml(text(r.building_name))}</td>` +
    `<td data-sort-value="${a}">${a}</td>` +
    `<td data-sort-value="${units}">${units}</td>` +
    landlordCell +
    `<td data-sort-value="${year}">${year}</td>` +
    `<td data-sort-value="${nIssues}">${nIssues}</td>` +
    `</tr>`
  );
}

export function buildingRowsHtml(records: BuildingRecord[]): string {
  return [...records]
    .sort((x, y) => descMissingLast(num(x.units), num(y.units)))
    .map(buildingRow)
    .join('\n');
}

const avgUnits = (total: number, bldgs: number): number => (bldgs > 0 ? roundHalfEven(total / bldgs, 1) : 0);

export function blockRowsHtml(fc: BlocksCollection): string {
  const rows = fc.features.map((f) => f.properties ?? ({} as BlocksCollection['features'][number]['properties']));
  return rows
    .map((p) => ({ p, label: text(p.block_label) }))
    .sort((x, y) => {
      const ux = x.label.startsWith('(Unknown)') ? 1 : 0;
      const uy = y.label.startsWith('(Unknown)') ? 1 : 0;
      return ux - uy || byCodePoint(x.label, y.label);
    })
    .map(({ p, label }) => {
      const id = Math.trunc(Number(p.block_id));
      const bldgs = Math.trunc(num(p.buildings) ?? 0);
      const totalRaw = num(p.total_units) ?? 0;
      const total = Math.trunc(totalRaw);
      const avg = avgUnits(totalRaw, bldgs);
      const median = num(p.median_year_built);
      const year = median === null ? '' : String(roundHalfEven(median));
      const l = escapeHtml(label);
      const chinatown = p.in_chinatown === true ? '1' : '';
      const village = p.in_village_plan === true ? '1' : '';
      return (
        `<tr data-block="${id}" data-area="${escapeHtml(text(p.local_area))}" ` +
        `data-chinatown="${chinatown}" data-village="${village}" ` +
        `data-bldgs="${bldgs}" data-units="${total}">` +
        `<td class="select-cell"><input type="checkbox" class="row-select" ` +
        `data-type="block" data-target="${id}"></td>` +
        `<td data-sort-value="${l}">${l}</td>` +
        `<td data-sort-value="${bldgs}">${bldgs}</td>` +
        `<td data-sort-value="${total}">${total}</td>` +
        `<td data-sort-value="${avg}">${avg.toFixed(1)}</td>` +
        `<td data-sort-value="${year}">${year}</td>` +
        `</tr>`
      );
    })
    .join('\n');
}

interface Group {
  label: string;
  key: string;
  buildings: number;
  totalUnits: number;
  records: BuildingRecord[];
}

/**
 * Group by key (count address, sum units), then sort by units, buildings desc.
 * One row per key: spellings of the same key ("Chartwell Construction Ltd" /
 * "CHARTWELL CONSTRUCTION LTD.") merge, labelled by the commonest spelling
 * (ties: code-point order), since wiring.js indexes markers by key alone.
 */
function aggregate(records: BuildingRecord[], labelOf: (r: BuildingRecord) => string, keyOf: (r: BuildingRecord) => string): Group[] {
  const groups = new Map<string, Group & { labels: Map<string, number> }>();
  for (const r of records) {
    const label = labelOf(r);
    const key = keyOf(r);
    let g = groups.get(key);
    if (!g) {
      g = { label, key, buildings: 0, totalUnits: 0, records: [], labels: new Map() };
      groups.set(key, g);
    }
    g.records.push(r);
    g.labels.set(label, (g.labels.get(label) ?? 0) + 1);
    if (!isMissing(r.address)) g.buildings += 1;
    g.totalUnits += num(r.units) ?? 0;
  }
  for (const g of groups.values()) {
    g.label = [...g.labels].sort((a, b) => b[1] - a[1] || byCodePoint(a[0], b[0]))[0][0];
  }
  return [...groups.values()]
    .sort((a, b) => byCodePoint(a.label, b.label) || byCodePoint(a.key, b.key)) // groupby's key order
    .sort((a, b) => b.totalUnits - a.totalUnits || b.buildings - a.buildings);
}

function groupRow(
  g: Group, type: 'owner' | 'network' | 'neighbourhood', target: string, first: string, dataKey: string, icon = '',
  pinned = false,
): string {
  const units = Math.trunc(g.totalUnits);
  const avg = avgUnits(g.totalUnits, g.buildings);
  const pin = pinned ? 'data-pin="last" class="not-on-record" ' : '';
  const label = pinned ? 'Not on record' : first;
  return (
    `<tr ${dataKey}="${target}" ${pin}` +
    `data-bldgs="${g.buildings}" data-units="${units}">` +
    `<td class="select-cell"><input type="checkbox" class="row-select" ` +
    `data-type="${type}" data-target="${target}"></td>` +
    `<td data-sort-value="${pinned ? '' : first}">${label}${icon}</td>` +
    `<td data-sort-value="${g.buildings}">${g.buildings}</td>` +
    `<td data-sort-value="${units}">${units}</td>` +
    `<td data-sort-value="${avg}">${avg.toFixed(1)}</td>` +
    `</tr>`
  );
}

/** Landlord groups with the "Not on record" group moved to the end. */
function landlordGroups(records: BuildingRecord[], nameOf: (r: BuildingRecord) => unknown, keyOf: (r: BuildingRecord) => unknown): Group[] {
  const groups = aggregate(
    records,
    (r) => (notOnRecord(keyOf(r)) ? '' : text(nameOf(r))),
    (r) => (notOnRecord(keyOf(r)) ? 'unknown' : String(keyOf(r))),
  );
  return [...groups.filter((g) => g.key !== 'unknown'), ...groups.filter((g) => g.key === 'unknown')];
}

/** A row's provenance icon, or '' when there is nothing to say. */
const iconFor = (text: string) => (text ? provenanceIcon(text) : '');

export function ownerRowsHtml(records: BuildingRecord[], licenceYear: number | null = null): string {
  return landlordGroups(records, (r) => r.owner_name, (r) => r.owner_key)
    .map((g) => {
      if (g.key === 'unknown') return groupRow(g, 'owner', 'unknown', '', 'data-owner', '', true);
      return groupRow(
        g, 'owner', escapeHtml(g.key), escapeHtml(g.label), 'data-owner',
        iconFor(ownerRowProvenance(g.records.map((r) => r.owner_source), licenceYear)),
      );
    })
    .join('\n');
}

export function networkRowsHtml(records: BuildingRecord[]): string {
  return landlordGroups(records, (r) => r.network_name, (r) => r.network_key)
    .map((g) => {
      if (g.key === 'unknown') return groupRow(g, 'network', 'unknown', '', 'data-network', '', true);
      const sample = g.records[0];
      const icon = sample.network_source === 'claims' || sample.network_source === 'licence'
        ? provenanceIcon(networkProvenance(sample))
        : '';
      const landlords = new Set(g.records.map((r) => r.owner_key).filter((k) => !notOnRecord(k))).size;
      const members = landlords > 1 ? `<span class="cell-sub-inline"> · ${landlords} landlords</span>` : '';
      return groupRow(g, 'network', escapeHtml(g.key), escapeHtml(g.label), 'data-network', members + icon);
    })
    .join('\n');
}

export function neighbourhoodRowsHtml(records: BuildingRecord[]): string {
  return aggregate(
    records,
    (r) => (isMissing(r.local_area) ? '(Unknown)' : String(r.local_area)),
    (r) => (isMissing(r.local_area) ? '(Unknown)' : String(r.local_area)),
  )
    .map((g) => {
      const area = escapeHtml(g.label);
      return groupRow(g, 'neighbourhood', area, area, 'data-area');
    })
    .join('\n');
}

/**
 * Adds the explainer icon to every header marked data-explain="owner|network".
 * Header cells are sticky (so they contain the tooltip) and tables can be wider
 * than the screen, so the tooltip is placed inside the visible table area each
 * time it opens.
 */
export function addHeaderExplainers(doc: Document = document): void {
  doc.querySelectorAll<HTMLElement>('table.data th[data-explain]').forEach((th) => {
    const text = EXPLAINERS[th.dataset.explain as keyof typeof EXPLAINERS];
    if (!text || th.querySelector('.prov')) return;
    th.insertAdjacentHTML('beforeend', provenanceIcon(text, 'About'));
    const icon = th.querySelector<HTMLElement>('.prov');
    if (!icon) return;
    const place = () => {
      const clip = (icon.closest('.table-wrap') ?? doc.documentElement).getBoundingClientRect();
      const viewWidth = doc.defaultView?.innerWidth ?? clip.right;
      const { left, width } = tipPlacement(
        icon.getBoundingClientRect().left, th.getBoundingClientRect().left,
        Math.max(clip.left, 0), Math.min(clip.right, viewWidth),
      );
      icon.style.setProperty('--tip-left', `${left}px`);
      icon.style.setProperty('--tip-width', `${width}px`);
    };
    icon.addEventListener('mouseenter', place);
    icon.addEventListener('focus', place);
  });
}

export function renderTables(
  data: BuildingData, blocks: BlocksCollection, doc: Document = document, licenceYear: number | null = null,
): void {
  const records = Object.values(data.records);
  const fill = (tableId: string, html: string) => {
    const tbody = doc.querySelector(`#${tableId} tbody`);
    if (tbody) tbody.innerHTML = html;
  };
  fill('buildings-table', buildingRowsHtml(records));
  fill('blocks-table', blockRowsHtml(blocks));
  fill('owners-table', ownerRowsHtml(records, licenceYear));
  fill('networks-table', networkRowsHtml(records));
  fill('neighbourhoods-table', neighbourhoodRowsHtml(records));
  addHeaderExplainers(doc);
}
